metaserver/src/main/scala/scala/meta/languageserver/storage/LevelDBMap.scala

```diff
@@ -0,0 +1,93 @@
+package scala.meta.languageserver.storage
+
+import java.io.File
+import com.typesafe.scalalogging.LazyLogging
+import monix.execution.schedulers.TestScheduler.Task
+import monix.reactive.Observable
+import org.fusesource.leveldbjni.JniDBFactory
+import org.iq80.leveldb.DB
+import org.iq80.leveldb.DBException
+import org.iq80.leveldb.Options
+
+/**
+ * A Scala-friendly wrapper around the JniDBFactory Java-wrapper around leveldb.
+ *
+ * @param db The leveldb, remember to close it after using. This wrapper will NOT
+ *           close the db for you.
+ */
+class LevelDBMap(db: DB) extends LazyLogging {
+
+  /** Returns the value matching key, if any. */
+  @throws[DBException]
+  def get[Key, Value](key: Key)(
+      implicit
+      keys: ToBytes[Key],
+      values: FromBytes[Value]
+  ): Option[Value] = {
+    Option(db.get(keys.toBytes(key))).map(values.fromBytes)
+  }
+
+  /**
+   * Gets the value if it exists, otherwise computes the fallback value and stores it.
+   *
+   * This method is not thread-safe, the computed fallback value may get overwritten.
+   */
+  @throws[DBException]
+  def getOrElseUpdate[Key, Value](key: Key, orElse: () => Value)(
+      implicit
+      keys: ToBytes[Key],
+      valuesFrom: FromBytes[Value],
+      valuesTo: ToBytes[Value]
+  ): Value = {
+    get(key) match {
+      case Some(value) => value
+      case None =>
+        val computed = orElse()
+        put(key, computed)
+    }
+  }
+
+  /** Inserts a new value for the given key. */
+  @throws[DBException]
+  def put[Key, Value](key: Key, value: Value)(
+      implicit
+      keys: ToBytes[Key],
+      values: ToBytes[Value]
+  ): Value = {
+    db.put(keys.toBytes(key), values.toBytes(value))
+    value
+  }
+
+  def close(): Unit = db.close()
+}
+
+object LevelDBMap {
+
+  /** Construct new wrapper around a leveldb. */
+  def apply(db: DB): LevelDBMap =
+    new LevelDBMap(db)
+
+  /**
+   * Creates a new leveldb in the given directory.
+   *
+   * Make sure to `db.close()`.
+   */
+  def createDBThatIPromiseToClose(directory: File): DB = {
+    val options = new Options
+    options.createIfMissing(true)
+    options.maxOpenFiles()
+    JniDBFactory.factory.open(directory, options)
+  }
+
+  def withDB[T](directory: File)(f: LevelDBMap => T): T = {
```

Comment:
Given that withDB looks safe and convenient, and createDBThatIPromiseToClose looks less attractive, when would I use the unsafe version?  Is there enough overhead to withDB that I don't want to use it in some scenarios?

Reply:
I'm not sure yet. To be honest I'm kind of leaning towards sqlite with a single table

```
create table index(
key bytes primary key,
value bytes,
binaryVersion string
)
```
where we can upgrade binary version when we change the indexing algorithms and want clean up old indices.

It wouldn't be a big change from what we have now using plain jdbc, and we could avoid the locking issues.